server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java

```diff
@@ -115,67 +121,120 @@ public CompletionStage<Void> restore(ZipFile zip) {
-            CommandsFactory commandsFactory = cr.getCommandsFactory();
-            KeyPartitioner keyPartitioner = cr.getComponent(KeyPartitioner.class);
-            InvocationHelper invocationHelper = cr.getComponent(InvocationHelper.class);
-            StorageConfigurationManager scm = cr.getComponent(StorageConfigurationManager.class);
-            PersistenceMarshaller persistenceMarshaller = cr.getPersistenceMarshaller();
-            Marshaller userMarshaller = persistenceMarshaller.getUserMarshaller();
-
-            boolean keyMarshalling = !scm.getKeyStorageMediaType().isBinary();
-            boolean valueMarshalling = !scm.getValueStorageMediaType().isBinary();
-
-            SerializationContextRegistry ctxRegistry = SecurityActions.getGlobalComponentRegistry(cm).getComponent(SerializationContextRegistry.class);
-            ImmutableSerializationContext serCtx = ctxRegistry.getPersistenceCtx();
-
-            int entries = 0;
-            try (DataInputStream is = new DataInputStream(zip.getInputStream(zipEntry))) {
-               while (is.available() > 0) {
-                  CacheBackupEntry entry = readMessageStream(serCtx, CacheBackupEntry.class, is);
+         int batchSize = (int) (Long.highestOneBit(ProcessorInfo.availableProcessors()) << 1);
+         log.debugf("Cache %s has file of size %d, batch size %d", cacheName, is.available(), batchSize);
+         Single<Long> restored = Flowable.fromIterable(() -> backupEntries)
+               .rebatchRequests(batchSize)
+               .map(entry -> {
                   Object key = keyMarshalling ? unmarshall(entry.key, userMarshaller) : scm.getKeyWrapper().wrap(entry.key);
                   Object value = valueMarshalling ? unmarshall(entry.value, userMarshaller) : scm.getValueWrapper().wrap(entry.value);
                   Metadata metadata = unmarshall(entry.metadata, persistenceMarshaller);
                   Metadata internalMetadataImpl = new InternalMetadataImpl(metadata, entry.created, entry.lastUsed);
 
                   PutKeyValueCommand cmd = commandsFactory.buildPutKeyValueCommand(key, value, keyPartitioner.getSegment(key),
                         internalMetadataImpl, FlagBitSets.IGNORE_RETURN_VALUES);
+                  commandsFactory.buildPutKeyValueCommand(key, value, keyPartitioner.getSegment(key),
+                        internalMetadataImpl, FlagBitSets.IGNORE_RETURN_VALUES);
                   cmd.setInternalMetadata(entry.internalMetadata);
-                  invocationHelper.invoke(cmd, 1);
-                  entries++;
-               }
-            } catch (IOException e) {
-               throw new CacheException(e);
-            }
-            log.debugf("Cache %s restored %d entries", cacheName, entries);
-         }, "restore-cache-" + cacheName));
+                  return cmd;
+               })
+               .flatMap(cmd -> {
+                  // Flowable does not accept null values.
```

Comment:
This should have used flatMapCompletable then you don't need the extra thenApply. The other option is to use RxJavaInterop.voidCompletionStageToFlowable which checks for some extra things and uses a much lighter conversion Processor as necessary.